src/index.js

```diff
@@ -11,30 +13,53 @@ ffmpeg.setFfprobePath(ffprobe.path);
 /**
  * @async
  * @param {string} url
- * @param {number} timeout
- * @returns {object} the ffprobe metadatra
+ * @param {object} opts request options (ie. `{ timeout: 1500 }`).
+ * @returns {object} the ffprobe metadata
  */
-module.exports = async function (url, timeout = 1500) {
-  return new Promise(function (resolve, reject) {
-    try {
-      const buffer = new PassThrough();
-      const stream = request.get(url, { timeout })
-        .on('error', (err) => {
-          buffer.end(); // close the buffer stream so everything finishes.
-          reject(err);
-        })
-        .pipe(buffer);
+module.exports = async function probe (url, opts = {}) {
+  const stream = got.stream(url, opts);
+  const input = opts.download ? await downloadStream(url, stream) : stream;
+  try {
+    const data = await _probe(input);
+    if (opts.download || (data && data.streams[0] && data.streams[0].profile !== 'unknown')) return data;
+    throw new Error('streaming probe failed, download and try again.');
+  } catch (err) {
+    return probe(url, { ...opts, download: true });
+  }
+};
 
-      ffmpeg()
-        .on('error', reject)
-        .input(stream)
-        // .ffprobe(streamIndex, callback)
-        .ffprobe(0, function (err, metadata) {
-          if (err) return reject(err);
-          resolve(metadata);
-        });
-    } catch (err) {
-      reject(err);
-    }
+/**
+ * execute the actual ffprobe call.
+ *
+ * @private
+ * @async
+ * @param {string|stream} input either a stream or a filename.
+ * @returns {object} the ffprobe metadata
+ */
+async function _probe (input) {
+  return new Promise((resolve, reject) => {
+    ffmpeg()
+      .on('error', reject)
+      .input(input)
+      .ffprobe(0, (err, data) => err ? reject(err) : resolve(data));
   });
-};
+}
+
+/**
+ * download a stream to a temp file.
+ *
+ * @private
+ * @async
+ * @param {stream} stream
+ * @returns {string} filename
+ */
+async function downloadStream (url, stream) {
```

Comment:
@aeh I already wrote a module that might do exactly what we need here... **https://github.com/noblesamurai/retriable-download**

Could save some LOC and keep maintenance lower?